aimmo-game/simulation/action.py

```diff
@@ -1,75 +1,154 @@
 from logging import getLogger
 from simulation.direction import Direction
 from simulation.event import FailedAttackEvent, FailedMoveEvent, MovedEvent, PerformedAttackEvent, ReceivedAttackEvent
-import simulation.world_map as world_map_module
 
 LOGGER = getLogger(__name__)
 
 
 class Action(object):
-    def apply(self, game_state, avatar):
+    def __init__(self, avatar):
+        self._avatar = avatar
+        try:
+            self._target_location = self._avatar.location + self.direction
+        except AttributeError:
+            self._target_location = self._avatar.location
+
+    @property
+    def avatar(self):
+        return self._avatar
+
+    @property
+    def target_location(self):
+        return self._target_location
+
+    def register(self, world_map):
+        if world_map.is_on_map(self.target_location):
+            world_map.get_cell(self.target_location).actions.append(self)
+
+    def process(self, world_map):
+        if self.is_legal(world_map):
+            self.apply(world_map)
+        else:
+            self.reject()
+
+    def is_legal(self, world_map):
+        raise NotImplementedError('Abstract method')
+
+    def apply(self, world_map):
+        raise NotImplementedError('Abstract method')
+
+    def reject(self):
         raise NotImplementedError('Abstract method')
 
 
 class WaitAction(Action):
-    def apply(self, game_state, avatar):
-        _add_score_from_cell_if_needed(avatar, game_state)
+    def __init__(self, avatar):
+        super(WaitAction, self).__init__(avatar)
+
+    def is_legal(self, world_map):
+        return True
+
+    def apply(self, world_map):
+        self.avatar.clear_action()
```

Comment:
it should never be called since `is_legal` is defined as `true`, so I think the `NotImplementedError` is desirable if it were ever called